### PR TITLE
Update Id::try_from to return a u32 and remove redundant logic in nut13::derive_path_from_keyset_id

### PR DESCRIPTION
**Pull request by vnprc**
Recreating [this PR ](https://github.com/cashubtc/cdk/pull/447)without including other contributors' changes in the diff.

The existing `TryFrom<Id>` trait implementation returns a u64 but can only contain values that fit within a u32 according to the [nut13 protocol definition](https://github.com/cashubtc/nuts/blob/main/13.md#keyset-id). This PR changes the existing implementation to return a u32 and defines two new TryFrom trait implementations that can be used to convert a keyset ID to and from a u64 with no data loss.

I have also opened a PR to clarify the language in NUT13 [here](https://github.com/cashubtc/nuts/pull/189).

### CODE REVIEW

File: crates/cdk/src/nuts/nut02.rs
Comment (thesimplekid) [inline crates/cdk/src/nuts/nut02.rs:143]: I'm hesitant to put in these u64 conversion and have the u32 one that does the module as defined in the nut. I think it maybe best to change the current `TryFrom<Id> for u64` to `TryFrom<Id> for u32` as you've done but to leave the u64 conversion out. Converting to and from u64 can be done outside cdk if thats needed for something as cdk should be kept to whats defined in the nuts.

Comment (vnprc) [reply crates/cdk/src/nuts/nut02.rs]: I see now I broke some other stuff with this change to u32. It's definitely good to isolate this change.

Comment (vnprc) [reply crates/cdk/src/nuts/nut02.rs]: don't merge, nut13 unit test incoming...

**Comment by vnprc** [comment]
I wrote a test for the nut13 functionality and ran it on the commit where it was broken...and it passed! Turns out we were reducing the keyset ID to 32 bits twice, once in the TryFrom impl and again in `derive_path_from_keyset_id`. Now with a unit test we can be sure that deriving the path works correctly and only reduce the keyset to u32 once.

PS I'm having to use `--no-verify` to commit. Not sure how to fix it but I can run all the checks using just.

**Comment by thesimplekid** [comment]
> Turns out we were reducing the keyset ID to 32 bits twice, once in the TryFrom impl and again in derive_path_from_keyset_id

Yeah I noticed that, I'm gonna write some full integration tests deriving from seed to make sure we haven't changed that before we merge this. 

> Not sure I agree with you about leaving out the u64 TryFrom impl but you are the captain. Disagree and commit. 🫡

Happy to hear a counter argument. 

> PS I'm having to use --no-verify to commit. Not sure how to fix it but I can run all the checks using just.

I know precommit still giving issues going to fix that here https://github.com/cashubtc/cdk/pull/451

**Comment by thesimplekid** [comment]
Noticed while we're here we can actually make it a `From` instead of a `TryFrom` if we add that to https://github.com/cashubtc/cdk/blob/f3ae4f4862ef42196fa24324750655175763cd7d/crates/cdk/src/nuts/nut02.rs#L107 can do this as a different pr if you don't want to include it here. 

```rust
    pub fn as_bytes(&self) -> [u8; Self::BYTELEN + 1] {
        let mut bytes = [0u8; Self::BYTELEN + 1];
        bytes[0] = self.version.to_byte();
        bytes[1..].copy_from_slice(&self.id);
        bytes
    }
```

**Comment by vnprc** [comment]
Nice!

**Comment by thesimplekid** [comment]
> Another potential reason not to implement From<Id> for u64 and it's inverse is that it could create confusion with From<Id> for u32. This is best solved with code documentation, IMO.

This is my issue with it - not that it's not in the spec but that it's potentially in conflict with what is. If someone has a u32 they may expect they can safely convert it to u64 to do From<u64> for Id, but that conversion path isn't actually safe since they can't know if data was lost when converting from Id to u32 via From<Id> for u32. Having these functions is too likely to lead to potential bugs, and adding docs isn't a reliable safeguard as documentation can be easily overlooked.

While I understand the desire to reduce complexity for higher-level developers and that is the goal of cdk, in this case the risk of subtle bugs outweighs the benefit. Even though implementing this trivial conversion outside cdk requires more work, it's preferable to having potentially unsafe implicit conversions in the library itself. This is especially important for a low-level library where predictability and safety should take precedence over convenience.

**Review by thesimplekid — APPROVED** [review]

**Comment by vnprc** [comment]
Gotcha, makes sense.

> If someone has a u32 they may expect they can safely convert it to u64 to do From for Id

Yeah I thought that was where you were gonna go initially. What do you think about this?

```
impl From<u32> for Id {
    fn from(value: Id) -> Self {
        panic!("Conversion from u32 to Id is not allowed. This operation is invalid and should never be performed.");
    }
}
```

**Comment by thesimplekid** [comment]
In general, we don't use panic in cdk other then tests always return an error. 

Here specifically, I don't think there is any need to add a function that will always panic or return an error the omission of a function supporting it implicitly signals it can't/shouldn't be done. This also introduces a runtime panic instead of a developer trying to do the conversion and realizing there is no function for it when developing a much preferable time for this realization. 

If we want to be more explicit about it being a one way function and why we can add to that documentation on the `From<Id> for u32`

**Pull request merged**
